app/src/main/java/com/arctouch/codechallenge/ui/custom/SearchAnimationToolbar.kt

```diff
@@ -0,0 +1,301 @@
+        closeButton.setImageResource(R.drawable.ic_close_grey_24dp)
+
+        // set hint and the text colors
+        txtSearch = searchView.findViewById<View>(android.support.v7.appcompat.R.id.search_src_text) as EditText
+        txtSearch.addTextChangedListener(this@SearchAnimationToolbar)
+        txtSearch.setOnEditorActionListener(TextView.OnEditorActionListener { _, actionId, _ ->
+            if (actionId == EditorInfo.IME_ACTION_SEARCH || actionId == EditorInfo.IME_ACTION_DONE) {
+                notifySearchSubmitted(txtSearch.text.toString())
+                return@OnEditorActionListener true
+            }
+            false
+        })
+
+    }
+
+    override fun onDetachedFromWindow() {
+        txtSearch.removeTextChangedListener(this@SearchAnimationToolbar)
+        super.onDetachedFromWindow()
+    }
+
+    fun setSupportActionBar(act: AppCompatActivity) {
+        act.setSupportActionBar(toolbar)
+    }
+
+    fun onSearchIconClick(): Boolean {
+        expand()
+        searchMenuItem.expandActionView()
+        return true
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private fun circleReveal(isShow: Boolean) {
+
+        var width = toolbar.width
+        width -= resources.getDimensionPixelSize(R.dimen.action_button_min_width) / 2
+
+        val cx = width
+        val cy = toolbar.height / 2
+
+        val anim: Animator
+        anim = if (isShow)
+            ViewAnimationUtils.createCircularReveal(searchToolbar, cx, cy, 0f, width.toFloat())
+        else
+            ViewAnimationUtils.createCircularReveal(searchToolbar, cx, cy, width.toFloat(), 0f)
+
+        anim.duration = 250L
+
+        // make the view invisible when the animation is done
+        anim.addListener(object : AnimatorListenerAdapter() {
+            override fun onAnimationEnd(animation: Animator) {
+                if (!isShow) {
+                    super.onAnimationEnd(animation)
+                    searchToolbar.visibility = View.INVISIBLE
+                }
+            }
+        })
+
+        // make the view visible and start the animation
+        if (isShow) {
+            searchToolbar.visibility = View.VISIBLE
+            notifySearchExpanded()
+        } else {
+            notifySearchCollapsed()
+        }
+
+        // start the animation
+        anim.start()
+    }
+
+    fun onBackPressed(): Boolean {
+        val isInSearchMode = searchToolbar.visibility == View.VISIBLE
+        if (!isInSearchMode) {
+            return false
+        }
+        collapse()
+        searchMenuItem.collapseActionView()
+        return true
+    }
+
+    fun setOnSearchQueryChangedListener(onSearchQueryChangedListener: OnSearchQueryChangedListener) {
+        this.onSearchQueryChangedListener = onSearchQueryChangedListener
+    }
+
+    private fun notifySearchCollapsed() {
+        if (this.onSearchQueryChangedListener != null) {
+            this.onSearchQueryChangedListener!!.onSearchCollapsed()
```

Comment:
Avoid using `!!` which is a Kotlin operator to force a non-null property to be null. Instead of it, you can overcome null checks using null-safety operator and `let` function, like the code excerpt below:

    this.onSearchQueryChangedListener?.let { listener ->
        listener.onSearchCollapsed()
    }

In that way, the listener will only be called when it is not null.

Reply:
Yeh or only using:
 `this.onSearchQueryChangedListener?.onSearchCollapsed()` 
This class was a Java class that I have copied from another project, I converted to Kotlin and didn't revise the details